### PR TITLE
ARROW-13830: [C++][Gandiva] Implement CHR Hive functions on Gandiva

### PR DESCRIPTION
**Pull request by jpedroantunes**

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -1361,6 +1361,26 @@ gdv_int32 ascii_utf8(const char* data, gdv_int32 data_len) {
   return static_cast<gdv_int32>(data[0]);
 }
 
+// Returns the ASCII character having the binary equivalent to A.
+// If A is larger than 256 the result is equivalent to chr(A % 256).
+GANDIVA_EXPORT
+const char* chr_int32(gdv_int64 context, gdv_int32 in, gdv_int32* out_len) {
+  in = in > 255 ? in % 256 : in;
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops.cc:1368]: keep it as in = in % 256;
i checked hive chr also prints for negative numbers

File: cpp/src/gandiva/precompiled/string_ops_test.cc [hunk 2]
```diff
@@ -53,6 +53,37 @@ TEST(TestStringOps, TestAscii) {
   EXPECT_EQ(ascii_utf8("999", 3), 57);
 }
 
+TEST(TestStringOps, TestChr) {
+  // CHR
+  gandiva::ExecutionContext ctx;
+  uint64_t ctx_ptr = reinterpret_cast<gdv_int64>(&ctx);
+  int32_t out_len = 0;
+
+  auto out = chr_int32(ctx_ptr, 88, &out_len);
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops_test.cc:62]: Also add unit tests for punctuations, non printable control codes, non-ascii number (>127)

**Review by projjal — CHANGES_REQUESTED** [review]

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 1 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Applied your suggestions

File: cpp/src/gandiva/precompiled/string_ops_test.cc (hunk 2 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops_test.cc:62]: Done, I added unit tests for this cases too

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 3]
```diff
@@ -1361,6 +1361,26 @@ gdv_int32 ascii_utf8(const char* data, gdv_int32 data_len) {
   return static_cast<gdv_int32>(data[0]);
 }
 
+// Returns the ASCII character having the binary equivalent to A.
+// If A is larger than 256 the result is equivalent to chr(A % 256).
+GANDIVA_EXPORT
+const char* chr_int32(gdv_int64 context, gdv_int32 in, gdv_int32* out_len) {
+  in = in % 256;
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops.cc:1368]: Please move this after the if-statement

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 4]
```diff
@@ -1361,6 +1361,26 @@ gdv_int32 ascii_utf8(const char* data, gdv_int32 data_len) {
...
+  if (in <= 0) {
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops.cc:1369]: What about 0? Why is this returning an empty string for 0? Hive UDF is not returning an empty string for 0

File: cpp/src/gandiva/precompiled/string_ops_test.cc (hunk 2 above)
Comment (vvellanki) [reply cpp/src/gandiva/precompiled/string_ops_test.cc:62]: Please add a test for 0 and 255

File: cpp/src/gandiva/precompiled/string_ops_test.cc [hunk 5]
```diff
@@ -53,6 +53,70 @@ TEST(TestStringOps, TestAscii) {
...
+  EXPECT_EQ(std::string(out, out_len), "T");
+
+  out = chr_int32(ctx_ptr, -5, &out_len);
+  EXPECT_EQ(std::string(out, out_len), "");
+
+  out = chr_int32(ctx_ptr, -340, &out_len);
+  EXPECT_EQ(std::string(out, out_len), "");
+
+  out = chr_int32(ctx_ptr, 256, &out_len);
+  EXPECT_EQ(std::string(out, out_len), "");
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops_test.cc:84]: This is wrong, it should return the char value of 0

File: cpp/src/gandiva/function_registry_string.cc [hunk 6]
```diff
@@ -65,6 +65,9 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
       UNARY_SAFE_NULL_NEVER_BOOL_FN(isnull, {}),
       UNARY_SAFE_NULL_NEVER_BOOL_FN(isnotnull, {}),
 
+      NativeFunction("chr", {}, DataTypeVector{int32()}, utf8(), kResultNullIfNull,
```
Comment (vvellanki) [inline cpp/src/gandiva/function_registry_string.cc:68]: Hive implements this for bigint and double. I dont see the Hive documentation implementing this for int

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 4 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc:1369]: It now returns the '\0' character which represents the NULL character

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 3 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc:1368]: moved

File: cpp/src/gandiva/precompiled/string_ops_test.cc (hunk 5 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops_test.cc:84]: It now returns the '\0' character which represents the NULL character

File: cpp/src/gandiva/function_registry_string.cc (hunk 6 above)
Comment (augustoasilva) [reply cpp/src/gandiva/function_registry_string.cc]: Changed this to bigint and double

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 7]
```diff
@@ -1361,6 +1361,48 @@ gdv_int32 ascii_utf8(const char* data, gdv_int32 data_len) {
   return static_cast<gdv_int32>(data[0]);
 }
 
+// Returns the ASCII character having the binary equivalent to A.
+// If A is larger than 256 the result is equivalent to chr(A % 256).
+FORCE_INLINE
+const char* chr_int64(gdv_int64 context, gdv_int64 in, gdv_int32* out_len) {
+  if (in < 0) {
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops.cc:1368]: Even for negative numbers, chr is returning valid values

For e.g. chr(-158) = 'b'

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 8]
```diff
@@ -1361,6 +1361,48 @@ gdv_int32 ascii_utf8(const char* data, gdv_int32 data_len) {
...
+    *out_len = 0;
+    return "";
+  }
+  in = (in > 256) ? in % 256 : in;
```
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops.cc:1372]: Please change this to:
in = in % 256

Comment (vvellanki) [reply cpp/src/gandiva/precompiled/string_ops.cc:1372]: Please add tests for negative numbers as input

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 7 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Change the behavior to handle negative number as well

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 8 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Changed and added test for negative number

**Review by vvellanki — APPROVED** [review]

**Review by vvellanki — APPROVED** [review]

**Pull request closed without merging**